### PR TITLE
Adding support for address lookup tables

### PR DESCRIPTION
**Pull request by godmodegalactus**

### CODE REVIEW

File: src/atl_store.rs [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    }
+
+    pub async fn load_atl_from_rpc(&self, atl: &Pubkey) {
+        if !self.map.contains_key(&atl.to_string()) {
+            self.reload_atl_from_rpc(&atl).await;
+        }
+    }
+
+    pub async fn reload_atl_from_rpc(&self, atl:&Pubkey) {
+        let response = self.rpc_client.get_account_with_commitment(atl, CommitmentConfig::processed()).await;
```
Comment (grooviegermanikus) [inline src/atl_store.rs:30]: shouldn't that be confirmed?

File: src/atl_store.rs [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+        if let Ok(account_res) = response {
+            if let Some(account) = account_res.value {
+                self.map.insert(atl.to_string(), account.data);
+            }
+        }
+    }
+
+    pub async fn load_accounts(
+        &self,
+        current_slot: Slot,
```
Comment (grooviegermanikus) [inline src/atl_store.rs:40]: "current" is misleading imho

File: src/atl_store.rs [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+impl ATLStore {
+    pub fn new(rpc_client: Arc<RpcClient>) -> Self {
+        Self {
+            rpc_client,
+            map: Arc::new(DashMap::new()),
+        }
+    }
+
+    pub async fn load_atl_from_rpc(&self, atl: &Pubkey) {
+        if !self.map.contains_key(&atl.to_string()) {
```
Comment (grooviegermanikus) [inline src/atl_store.rs:24]: can't you use pubkey type?

File: src/atl_store.rs [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+                let lookup_table = AddressLookupTable::deserialize(&account.value()).unwrap();
+                let write_accounts = lookup_table
+                    .lookup(current_slot, write_accounts, &SlotHashes::default());
+                let read_account = lookup_table
+                    .lookup(current_slot, read_account, &SlotHashes::default());
+
+                let write_accounts = if let Ok(write_accounts) = write_accounts {
+                    write_accounts
+                } else {
+                    return None;
```
Comment (grooviegermanikus) [inline src/atl_store.rs:56]: where is the return jumping to?

File: src/atl_store.rs [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+            None => {
+                Some(vec![])
+            }
+        }
+    }
+
+    pub async fn get_accounts(
+        &self,
+        current_slot: Slot,
+        atl: Pubkey,
```
Comment (grooviegermanikus) [inline src/atl_store.rs:93]: could be &Pubkey

File: src/main.rs [hunk 6]
```diff
@@ -57,16 +60,14 @@ pub async fn start_tracking_banking_stage_errors(
...
-                "slot_sub".to_string(),
+                "slot_sub_for_banking_tx".to_string(),
                 yellowstone_grpc_proto::geyser::SubscribeRequestFilterSlots {},
             );
             slot_sub
-        } else {
-            HashMap::new()
         };
 
         let mut geyser_stream = client
```
Comment (grooviegermanikus) [inline src/main.rs:73]: can use the grpc-geyser+connector lib

**Comment by grooviegermanikus** [comment]
is it ALT or ATL ?

File: migration.sql [hunk 7]
```diff
@@ -58,6 +58,8 @@ CREATE TABLE banking_stage_results_2.accounts_map_transaction(
   PRIMARY KEY (acc_id, transaction_id)
 );
 
+ALTER TABLE banking_stage_results_2.accounts_map_transaction ADD COLUMN is_atl BOOL;
```
Comment (grooviegermanikus) [inline migration.sql:61]: IMO we should start to append patches at the bottom at the file to make them applicable in sorted order

File: src/main.rs [hunk 8]
```diff
@@ -224,7 +241,15 @@ async fn start_tracking_blocks(
...
                     });
                     // delay queue so that we get all the banking stage errors before processing block
-                }
+                },
+                yellowstone_grpc_proto_original::prelude::subscribe_update::UpdateOneof::Account(account_update) => {
+                    info!("ATL updated");
+                    if let Some(account) = account_update.account {
+                        let bytes: [u8; 32] = account.pubkey.try_into().unwrap_or(Pubkey::default().to_bytes());
+                        let pubkey = Pubkey::new_from_array(bytes);
+                        atl_store.map.insert( pubkey.to_string(), account.data);
```
Comment (grooviegermanikus) [inline src/main.rs:250]: _map_ should not be accesses directly IMO

File: src/atl_store.rs [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+        write_accounts: &Vec<u8>,
+        read_account: &Vec<u8>,
+    ) -> Vec<TransactionAccount> {
+        self.load_atl_from_rpc(&atl).await;
+        match self.load_accounts(current_slot, atl, write_accounts, read_account).await {
```
Comment (grooviegermanikus) [inline src/atl_store.rs:98]: `load_accounts` should only to the deserialization;
`get_accounts` should rely on `load_atl_from_rpc` or something similar to get the data

File: src/atl_store.rs [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+use solana_address_lookup_table_program::state::AddressLookupTable;
+use solana_rpc_client::nonblocking::rpc_client::RpcClient;
+use solana_sdk::{pubkey::Pubkey, slot_hashes::SlotHashes, slot_history::Slot, commitment_config::CommitmentConfig};
+use std::sync::Arc;
+
+use crate::block_info::TransactionAccount;
+
+pub struct ATLStore {
+    rpc_client: Arc<RpcClient>,
+    pub map: Arc<DashMap<String, Vec<u8>>>,
```
Comment (grooviegermanikus) [inline src/atl_store.rs:12]: should control somehow the size of this map; maybe log the size if it exceeds a certain size

File: src/atl_store.rs (hunk 1 above)
Comment (godmodegalactus) [reply src/atl_store.rs:30]: Nah processed is correct.

File: src/atl_store.rs (hunk 4 above)
Comment (godmodegalactus) [reply src/atl_store.rs:56]: return from the function

File: src/atl_store.rs (hunk 10 above)
Comment (godmodegalactus) [reply src/atl_store.rs]: Nah should be ok for now. There are not many ATL accounts.

**Pull request merged**